DOC: Model tutorial format based on tutorial-svd

bjnath opened the pull request:
A proposed style and format for tutorials, using tutorial-svd as a basis.

Includes fixes for most points raised in issue #16244

Also includes edits intended to show up the quality of the tutorial.
Working through it, I could appreciate how much effort had
gone into it and how significantly it will benefit users.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

Commit:
ENH: improve printing of arrays with multi-line reprs

Commit:
Improve testing

Commit:
Improve way of breaking lines

Commit:
Make requested changes

WarrenWeckesser committed:
DEP: Deprecate `numpy.dual`.

Add a deprecation warning in the `numpy.dual` module, and
remove the use of `numpy.dual` from the few places where it
is used in the numpy code.

WarrenWeckesser committed:
DOC: dual: Note the deprecation of numpy.dual in the dual module docstring.

Also make the correction "Scipy" -> "SciPy" in a few places.

WarrenWeckesser committed:
REL: Add a release note about the deprecation of numpy.dual.

Commit:
Add release note snippet

timhoffm committed:
DOC: Unify cross-references between array joining methods

seberg committed:
MAINT: Unify cached (C-level static) imports

The main idea is that once we unify them, we can think about actually
being able to also clean them up.
This is far away from having no global state, but at this time I do not
see that as a seriousl goal personally. But it could be possible
to reload the module safely in a single interpreter maybe.

The commit adds the cache import for the complex casts, since it doubles
performance for small casts when the warning is not suppressed and it
unifies the handling.

WarrenWeckesser committed:
DOC: Warn about behavior of ptp with signed integers.

* Add a warning to the docstrings of numpy.ma.MaskedArray.ptp and
  numpy.ptp about a problem with signed integer arrays.
* Add an example to the the docstring of numpy.ptp that demonstrates
  the issue and suggests a work-around.
* Add an "Examples" section to the docstring of MaskedArray.ptp
  which includes the new example added to the docstring of numpy.ptp.
* Make the numbers in the basic example more interesting.

seberg committed:
BUG: Fix default fallback in genfromtxt

This affected (for example?) if the `dtype=object` was used
without a converter, meaning that the default one is used.
And this is currently the last one, which is `string_` (and thus
bytes).

Closes gh-16189

bjnath committed:
DOC: Model tutorial format based on tutorial-svd

A proposed style and format for tutorials, using tutorial-svd as a basis.

Includes fixes for most points raised in #16244

Also includes edits intended to show up the quality of the tutorial.
Working through it, I could appreciate how much effort had
gone into it and how significantly it will benefit users.

bjnath committed:
DOC: Model tutorial format based on tutorial-svd

A proposed style and format for tutorials, using tutorial-svd as a basis.

Includes fixes for most points raised in #

Also includes edits intended to show up the quality of the tutorial.
Working through it, I could appreciate how much effort had
gone into it and how significantly it will benefit users.

mattip committed:
Merge pull request #16240 from WarrenWeckesser/ptp-warning

DOC: Warn about behavior of ptp with signed integers.

seberg committed:
Merge pull request #16197 from timhoffm/doc-stack

DOC: Unify cross-references between array joining methods

seberg committed:
TST: Add a test for np.add_docstring

Its not quite the right file, but close to newdoc seemed sensible
and we do not have a "right" file right now...

WarrenWeckesser committed:
DOC: Update the f2py section of the "Using Python as Glue" page.

* Update the output shown for the docstrings generated by f2py.

rlintott committed:
BUG: numpy.einsum indexing arrays now accept numpy int type (gh-16080)

* Using PyArray_PyIntAsIntp helper function instead

* TST: add tests for einsum numpy int and bool list subscripts
Added tests to check that einsum accepts numpy int64 types and
rejects bool. Rejecting bools is new behaviour in subscript lists.
I changed ValueError to TypeError on line 2496 in multiarraymodule.c
as it is more appropriate. I also modified einsumfunc.py to have the
same behaviour as in the C file when checking subscript list.
(Reject bools but accept anything else from operator.index())

Closes gh-15961

sibasmarak committed:
DOC: Updated the f2py section of the "Using Python as Glue" page. (#16258)

* Replace print statements with print function calls.
* Fix the filename in one of the commands: it is "add.f", not "add.f95".

seberg committed:
BUG: Add missing decref in fromarray error path

This function steals the dtype, so it must decref it also
on error. This is fixup of ffe76ac8df65c8e7831df9924a782276e060f3e6
which accidentally deleted the decref.

charris committed:
ENH: Add tool for downloading release wheels from Anaconda.

This is a simplified version of terryfy::wheel-uploader that has
two advantages over the original:

- It works with Anaconda where our wheels are now stored.
- It is simplified to match the NumPY workflow.

jbrockmendel committed:
ENH: include dt64/td64 isinstance checks in __init__.pxd

charris committed:
Merge pull request #16263 from seberg/decref-fromarray-error-path-fixup

BUG: Add missing decref in fromarray error path

mattip committed:
Merge pull request #16265 from charris/add-download-wheel-tool

ENH: Add tool for downloading release wheels from Anaconda.

tbm committed:
DOC: Fix typos and cosmetic issues

mattip committed:
Merge pull request #16269 from tbm/typo

DOC: Fix typos and cosmetic issues

charris committed:
REL: Update master after 1.19.x branch.

- Delete release note fragments from 1.19.0 (towncrier)
- Update 1.19.x release note (towncrier)
- Create 1.20.0-notes.rst for master development
- Update C-API versions for 1.20.x
- Update setup.py for 1.20.0

eric-wieser committed:
MAINT: Remove a pointless if

The `add` ufunc is happy to handle `out=None` by itself

eric-wieser committed:
MAINT: Avoid moving axes around multiple times

It's easier to move the relevant axis to position 0 in `ap` first than it is to move it for every relevant object simultaneously.

eric-wieser committed:
MAINT: Extract a lerp helper function to make the algorithm of quantile clearer

This does not affect the behavior in any way

eric-wieser committed:
MAINT: Remove special cases for 0d arrays in quantile

This also simplifies the axis handling logic, taking advantage of the fact we know `axis` is 0 for the rest of the function body

eric-wieser committed:
BUG: Ensure out argument is returned by identity for 0d arrays

This makes the following functions consistent with the behavior of ufuncs:

* `np.argmax`
* `np.argmin`
* `np.choose`
* `np.trace`
* `np.round`
* `np.take`

Previously they would unpack these into scalars.

charris committed:
Merge pull request #16275 from charris/update-after-1.19.x-branch

REL: Update master after 1.19.x branch.

SacredArrow committed:
DOC: Clarifications for np.std (#16267)

* DOC: Clarifications for np.std

* Line lengths fixes in np.std

* Swapped variable names in np.std

SacredArrow committed:
DOC: Clarifications for np.var

eric-wieser committed:
Merge pull request #16274 from eric-wieser/extract-lerp

MAINT: cleanups to quantile

WarrenWeckesser committed:
DOC: A few small changes in the "Using Python as Glue" page.

* Fix whitespace in the inputs to Python in the f2py examples.
* Capitalize Python and Fortran consistently.

mattip committed:
Merge pull request #16238 from seberg/maint-cached-imports

MAINT: Unify cached (C-level static) imports

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.12.0 to 5.14.0

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.12.0 to 5.14.0.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.12.0...hypothesis-python-5.14.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

seberg committed:
Merge pull request #16257 from WarrenWeckesser/doc-glue-f2py

DOC: Update the f2py section of the "Using Python as Glue" page.

seberg committed:
Merge pull request #16278 from SacredArrow/var-doc-edit

DOC: Clarifications for ``np.var``.

seberg committed:
Merge pull request #16285 from numpy/dependabot/pip/hypothesis-5.14.0

MAINT: Bump hypothesis from 5.12.0 to 5.14.0

rk-mlu committed:
DOC: Calrify tiny/xmin in finfo and machar (gh-16253)

This commit adds a note to the documentation of finfo clarifying that
the attribute `tiny` does not represent the smallest positive usable
number. Instead it refers to the smallest positive number with there
being no leading 0's in the mantisse (normalized floating point number).
Alternatively, `tiny` represents the smallest positive floating point
number with full precision. The smallest positive subnormal floating
point number is typically orders of magnitudes smaller, but has reduced
precision since leading 0's in the mantisse are used to allow smaller
exponents than indicated by `minexp`.

The commit also adds a brief clarification to the docstring of machar.

Closes #16252

Co-authored-by: Anirudh Subramanian <anirudh2290@apache.org>
Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

guilhermeleobas committed:
DOC: Add PyArray_ContiguousFromObject C docs (gh-16207)

Add documentation for `PyArray_ContiguousFromObject` similar to the other
macros wrapping `PyArray_FromAny`

Fixes #16196

charris committed:
BUG: Fix tools/download-wheels.py.

`tools/download-wheels` was downloading an html page rather than the binary
wheel file.

mattip committed:
Merge pull request #16292 from charris/fix-download-wheels

BUG: Fix tools/download-wheels.py.

jbrockmendel committed:
Re-order declarations

jbrockmendel committed:
Missing declarations

jbrockmendel committed:
move below int64_t declaration

jbrockmendel committed:
troubleshoot CI

lkdmttg7 committed:
MAINT: Chain exceptions in generate_umath.py (gh-16254)

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>
Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

seberg committed:
Update numpy/lib/_iotools.py

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris committed:
BUG: Require Python >= 3.6 in setup.py

Closes #16294.

seberg committed:
Merge pull request #16276 from eric-wieser/take-out

 BUG: Ensure out argument is returned by identity for 0d arrays

bje- committed:
DOC: Add math.isclose to See Also (#16283)

* DOC: Move math.isclose note to See Also.

mattip committed:
Merge pull request #16295 from charris/fix-missed-3.5

BUG: Require Python >= 3.6 in setup.py

mattip committed:
Merge pull request #16242 from seberg/genfromtxt-default-regression

BUG: Fix default fallback in genfromtxt

eric-wieser committed:
BUG: np.info does not show keyword-only arguments

Using inspect.signature instead of `np.compat.getargspec` solves this problem.

`inspect.signature` also handles stripping the `self` argument of methods for us.

eric-wieser committed:
MAINT: remove workaround for fixed issue

take now correctly returns `out`, even on 0d arrays

seberg committed:
Merge pull request #16156 from WarrenWeckesser/deprecate-dual

DEP: Deprecate `numpy.dual`.

Commit:
Move the snippet note

blackpastures committed:
MAINT: precompute log(2.0 * M_PI) in `random_loggam' (gh-16237)

Most compilers should optimize it, but it doesn't hurt to inline and has a better name
now.

seberg committed:
BUG: Allow attaching documentation twice in add_docstring

This is technically not a bug, but some IDEs and IPython have
autoreload magic which can mean that NumPy gets reloaded a second
time. This is not safe, but when it happens ignoring that an
identical docstring is already attached fixes the issue.

seberg committed:
DOC: Fix generic scalar docstrings to point to ndarray

Previously some of these were not set, because they were already
set on the C-level. Remove the duplicate message and replace it
instead with a forward to the ``ndarray`` attribute/method.

These docstrings are inherited by the actual scalars and thus the
"virtual class" note was misleading.

seberg committed:
MAINT: Simplify logic in add_docstring

The old pointers are now provided by the C-API and the macros
seemed a bit confusing since ``new`` appears from nowhere being
defined in the macro.

Carreau committed:
DOC: Fix documentation rendering,

the => seem to be extra and rst seem to not interprete backtick with
leading space as closing a directive, the the full line on text was put
in a html pre-tag when building the docs.

see current https://numpy.org/doc/stable/reference/generated/numpy.packbits.html

seberg committed:
Merge pull request #16305 from Carreau/doc-micro-fix

DOC: Fix packbits documentation rendering,

seberg committed:
Merge pull request #16284 from eric-wieser/extract-lerp

MAINT: Clean up the implementation of quantile

mattip committed:
BUG: relpath fails for different drives on windows (#16308)

* BUG: relpath fails for different drives on windows

* ENH: always use abspath

Commit:
Improve the snippet

embray committed:
BLD: Avoid "visibility attribute not supported" warning (gh-16288)

I found that when building the latest master branch on Cygwin, while testing #16246, that thousands of warnings were generated at build time like:

numpy/core/src/npysort/binsearch.c.src: In function ‘binsearch_left_bool’:
numpy/core/src/npysort/binsearch.c.src:82:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
Granted this is just a warning, so I don't think it's a serious issue.

It seems the test that was supposed to check for __attribute__ support was not working as expected. The #pragmas only take effect if I provide a function body--they are ignored for bare declarations. I don't know if that's by intent, or if it's a GCC issue. For reference:

$ gcc --version
gcc (GCC) 7.4.0

panpiort8 committed:
Update doc/release/upcoming_changes/15997.improvement.rst

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

mattip committed:
Merge pull request #16239 from seberg/reimport-do-not-set-docs

BUG,DOC: Allow attach docs twice but error if wrong

mattip committed:
Merge pull request #15997 from panpiort8/array_pretty_print

ENH: improve printing of arrays with multi-line reprs

rossbar committed:
MAINT: Bumpy numpydoc version.

Fixes problem with citation (reference) anchors in rendered docs.

Actual fix was in numpy/numpydoc@7c42883, but this bumps to latest
(unreleased) numpydoc version.

seberg committed:
MAINT: Stop Using PyEval_Call* and simplify some uses

Python 3.9 started to deprecate these calls and the PyObject_Call*
variants are equivalent or nicer in any case.

seberg committed:
TST: Add 3.9-dev to the travis matrix

Co-authored-by: Charles Harris <charlesr.harris@gmail.com>

charris committed:
Merge pull request #16316 from rossbar/maint/bump_numpydoc

MAINT: Bump numpydoc version

charris committed:
Merge pull request #16318 from seberg/maint-python39

MAINT: Stop Using PyEval_Call* and simplify some uses

tinaoberoi committed:
DOC: Correct documentation of ``__array__`` when used as output array. (#16130)

lschroefl committed:
DOC: link np.interp to SciPy's multidimensional interpolation functions (#14154)

brandtbucher committed:
BUG: Don't segfault on bad __len__ when assigning. (gh-16327)

When __getitem__ fails, assignment falls back on __iter__, which may not
have the same length as __len__, resulting in a segfault.
See gh-7264.

* BUG: Don't rely on __len__ for safe iteration.

Skip __len__ checking entirely, since this has no guarantees of being
correct. Instead, first convert to PySequence_Fast and use that length.
Also fixes a refleak when creating a tmp array fails.
See gh-7264.

* TST: Update test for related edge-case.

The previous fix more gracefully handles this edge case by skipping the
__len__ check. Rather than raising with an unhelpful error message, just
create the array with whatever elements C() actually yields.
See gh-7264.

kevmo committed:
DOC: Fix spelling typo - homogenous to homogeneous. (#16324)

jbrockmendel committed:
Fix np.dtype declaration

rossbar commented:
@bjnath thanks for the extensive look at the tutorial - I'm very keen to take a closer look at the proposed changes! 

I don't think looking at the diffs is going to be sufficient to truly get a sense for the changes, especially since their will be differences in the sphinx-rendered output that we won't see in looking at the source file alone (for instance, this already bit me when I looked at the changes in #16259 and missed the fact that chunks of text had been *moved* rather than *deleted*).

Of course, reviewers could build locally, but it might help speed up the review process if there were a rendered version already available online. If you're comfortable doing so, posting the rendered changes (on the gh-pages branch of your fork of numpy for example) might make it easier to get more eyes on the changes. Just a thought, but by no means a requirement!

lschroefl committed:
Update numpy/lib/function_base.py

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

seberg committed:
BUG: Fix refcounting in add_newdoc

this is a trivial refcounting fix, since this function is mostly
used only at startup, the small refcounting issue was not noticed
until the new tests were added.

eric-wieser commented:
> but it might help speed up the review process if there were a rendered version already available onlin

The circleci build page contains just that.

rossbar commented:
> The circleci build page contains just that.

Wow, thanks for this @eric-wieser , I'm a bit ashamed to admit that I never realized that the results of the build devdocs step were available via the "Artifacts" tab. This one little comment is going to do wonders for my workflow reviewing doc changes :pray:

jbrockmendel committed:
First stab at tests, try on the CI

seberg committed:
CI: Create a link for the circleCI artifact

rossbar commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -303,92 +237,86 @@ As expected, you should see something like
-``Sigma`` (which are the same as in ``s``) as zeros, keeping
-``U`` and ``Vt`` intact, and computing the product of these matrices
-as the approximation.
+Although ``s`` has 768 singular values, most of them are pretty small. To
+build a more economical approximation, it might make sense to use only the
+first 50 or so.
```
In both the original text an here, there seems to be a disconnect between the number of singular values to keep in the text (50) and the number used in the code example (10). Either is fine but I think it would be an improvement to have the text and the code sync up.

bjnath replied:
This isn't a bug, it's a feature. 50 seems like a reasonable bound from the graph;  and for fun the tutorial says let's set k to 10, a really low value, to see how much of the image we can still see. k is meant to be a variable and not a constant; the tutorial tells the reader to try different k.

It would be more a disconnect to point to the graph and talk about limiting to the first 10, and it would be less interesting to see the picture with 50 when it turns out so plausibly with only 10.

We can certainly add text to emphasize that we're intentionally picking a low k.

rossbar reviewed (commented):
First of thank you both @melissawm for putting together this tutorial and @bjnath for the extensive comments (both in #16244 and demonstrated here). Overall, I think the tutorial is great and look forward to seeing many more official NumPy tutorials!

I read through the tutorial with the [original](https://numpy.org/devdocs/user/tutorial-svd.html) and the [version from this PR](https://13310-908607-gh.circle-artifacts.com/0/doc/build/html/user/tutorial-svd.html) side-by-side to try to get a sense for the differences. Generally speaking, I thought both versions were great: the original tended to feel a little more thorough and verbose, while this one is more concise. In both cases I found the tutorial clear and compelling. A couple other general comments:

 - The original tutorial has more parentheticals/notes, whereas this PR removes some of these asides from the main narrative and collects them near the bottom in a **In practice** section. I think both approaches are equally valid - maybe using sphinx's footnote mechanism is a viable middle ground?
 - The term "matrix" is used a lot to describe 2D arrays - this is one piece of terminology that we should be very careful with in tutorials given, for example, the difference between element-wise and matrix multiplication and the various defaults/operators (this is explicitly called out at one point in the tutorial text).

Finally, there are certainly differences between the tone/style of the writing, but the important thing is that (in my opinion at least) they are both clear and focused. I'm not sure how one would "evaluate" something like this, but FWIW I think both versions are very effective at communicating technical concepts.

mattip committed:
Merge pull request #16332 from LSchroefl/link_SciPy_multidimensional_interpolation

DOC: link np.interp to SciPy's interpolation functions (closes #14154)

mattip committed:
ENH: run tests for __init__.pxd

rgommers committed:
Merge pull request #16333 from kevmo/homogeneous-16324

DOC: Fix spelling typo - homogenous to homogeneous. (#16324)

stefanv commented:
>     * The original tutorial has more parentheticals/notes, whereas this PR removes some of these asides from the main narrative and collects them near the bottom in a **In practice** section. I think both approaches are equally valid - maybe using sphinx's footnote mechanism is a viable middle ground?

@rossbar The changes here were motivated in detail in https://github.com/numpy/numpy/issues/16244  Taking that into consideration, do you think it's merely a difference in style, or do you think the arguments in that issue hold merit?  We need to evaluate them individually and see whether we agree or not.

mattip committed:
Merge pull request #16337 from seberg/circleci-artifact

CI: Create a link for the circleCI artifact

bjnath commented:
@rossbar, thank you for reviewing the changes patiently and carefully. Your technical observations are valuable and on point. I'm grateful for these.

I'm troubled that, through no fault of yours, we're missing the larger issue. NEP 44 sets strategy; now we need tactics.

While #16244 addresses this tutorial, and raises questions about it that are still unanswered, its underlying subject is tactics. PR #16245 is the appendix where, having gone through the philosophy, I show you how it reads.

#16244 won't sound right to everyone and I expect to lose arguments.

But unless we choose a direction -- any direction -- for documentation, I'm valueless here. It's not enough to draw up a list; now we need living examples. I can't begin to help you build unless we agree on a set of plans.

charris committed:
Merge pull request #16336 from seberg/trivial-refcountfix-add_docstring

BUG: Fix refcounting in add_newdoc

seberg committed:
BUG: Fix dtype leak in `PyArray_FromAny` error path

Also adds a test to bincount which will run into this path.
The leak can be triggered by using a reference count checker
on the test suit (e.g. pytest-leaks).

Closes gh-16339

wrzadkow committed:
MAINT: Remove f-strings in setup.py. (gh-16346)

Remove f-strings from setup.py to allow for an informative error message for python<3.6 users. Closes #16345.

mattip committed:
BUILD: Remove Accelerate support (#15759)

Remove support for Apple Accelerate, since it is buggy. A build error should
occur on most or all setups if linked against Accelerate. Test or import failures
should occur on setups where Accelerate is picked up dynamically.

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>

larsoner committed:
BUG: Indentation for docstrings

rossbar commented:
> The changes here were motivated in detail in #16244 Taking that into consideration, do you think it's merely a difference in style, or do you think the arguments in that issue hold merit? We need to evaluate them individually and see whether we agree or not.

Understood - I tried to keep my comments here as high-level as possible, but I agree I didn't do a very good job connecting back to enumerated points in the corresponding issue. I will remedy this ASAP.

In this particular case: I am not a trained technical writer, so I have nothing to offer but my opinion, but I personally don't mind the extra notes & parenthetical statements. To me, whether to include these is a trade-off between how much the reader has to keep in their head at once vs. getting a more broad picture of overall context. I have no opinion on whether one is objectively better than the other. I do think @bjnath's comment in #16244 about making use of the more specific sphinx directives (the general `admonition`, but there are also purpose-built subclasses like `hint`, `warning`, `important` etc.) to help visually indicate to the reader that whatever is in these boxes is "extra" info and not intended to be part of the focused narrative.

> While #16244 addresses this tutorial, and raises questions about it that are still unanswered, its underlying subject is tactics. PR #16245 is the appendix where, having gone through the philosophy, I show you how it reads.

As I mentioned, I didn't do a very good job of tying the comments here back to the issue which explicitly stated the main points. I will remedy this by commenting directly on the issue (keeping what I've learned from reading through the tutorial in mind).

bjnath commented:
@rossbar My comments were not directed at you *at all*, and I am very, very sorry if they appeared to be.

I'm not calling for another round of review -- exactly the opposite. I'm suggesting we (members of community who want to talk docs) set this PR down and instead throw darts at issue #16244. 

I think you did a great job, and tying your comments more tightly to #16244 is extra work for you and is not at all the outcome I'm hoping for.

rossbar commented:
> My comments were not directed at you at all, and I am very, very sorry if they appeared to be.

Of course not! Your comment above was very helpful - it doesn't make much sense to review the *appendix* without addressing the document proper :). I think your take is correct: concentrating conversation in #16244 rather than the review mechanism here would be much more constructive.

jbrockmendel committed:
blackify, troubleshoot tests

seberg committed:
Merge pull request #16298 from eric-wieser/fix-np.info

BUG: np.info does not show keyword-only arguments

anirudh2290 committed:
MAINT: Cleanup 'tools/download-wheels.py' (#16329)

* ENH: Create download dir if not present for download-wheels

* TST: Add tests for download-wheels

* MAINT: use exist_ok=True for os.makedirs in download-wheels

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

* MAINT: Remove test for download_wheels

* MAINT: Remove directory creation and error if dir not present

* MAINT: Print total files download and remove err for 0 search results

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

charris committed:
Merge pull request #16349 from larsoner/dep

BUG: Indentation for docstrings

charris committed:
Merge pull request #16348 from seberg/fromany-leak

BUG: Fix dtype leak in `PyArray_FromAny` error path

mattip committed:
Merge pull request #16266 from jbrockmendel/cyapi

ENH: include dt64/td64 structs and funcs and tests in __init__.pxd.

Qiyu8 committed:
DOC: Reconstruct Testing Guideline. (#16323)

* reconstruct testing doc

* Update doc/TESTS.rst.txt

* DOC: add docstring for numpy.test

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>
Co-authored-by: Matti Picus <matti.picus@gmail.com>

charris committed:
MAINT: Streamline download-wheels.

- Makes specifying name patterns simpler
- Makes name pattern reject dev versions unless specified.
- Makes progress printing aligned and more concise.

tylerjereddy committed:
Merge pull request #16362 from charris/update-download-wheels

MAINT: Streamline download-wheels.

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.14.0 to 5.15.1

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.14.0 to 5.15.1.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.14.0...hypothesis-python-5.15.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

charris committed:
Merge pull request #16368 from numpy/dependabot/pip/hypothesis-5.15.1

MAINT: Bump hypothesis from 5.14.0 to 5.15.1

dependabot-preview[bot] committed:
MAINT: Bump pytest-cov from 2.8.1 to 2.9.0

Bumps [pytest-cov](https://github.com/pytest-dev/pytest-cov) from 2.8.1 to 2.9.0.
- [Release notes](https://github.com/pytest-dev/pytest-cov/releases)
- [Changelog](https://github.com/pytest-dev/pytest-cov/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest-cov/compare/v2.8.1...v2.9.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
MAINT: Bump cython from 0.29.17 to 0.29.19

Bumps [cython](https://github.com/cython/cython) from 0.29.17 to 0.29.19.
- [Release notes](https://github.com/cython/cython/releases)
- [Changelog](https://github.com/cython/cython/blob/master/CHANGES.rst)
- [Commits](https://github.com/cython/cython/compare/0.29.17...0.29.19)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

charris committed:
Merge pull request #16367 from numpy/dependabot/pip/cython-0.29.19

MAINT: Bump cython from 0.29.17 to 0.29.19

charris committed:
Merge pull request #16369 from numpy/dependabot/pip/pytest-cov-2.9.0

MAINT: Bump pytest-cov from 2.8.1 to 2.9.0

melissawm commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -1,6 +1,6 @@
-Tutorial: Linear algebra on n-dimensional arrays
-================================================
+===============================================================
+Tutorial: A tour of n-dimensional arrays
```
+1 for tags! I also agree that the original title was more specific, but I'll let others have their say.

bjnath replied:
The original title doesn't sound like a tutorial title -- it sounds like the title of an explanation page. 

I tried out a title like "Compress an image using SVD", but while that's factually true, it doesn't reflect the spirit of the tutorial; SVD is an interesting way to get the user's hands on n-dimensional arrays.

It could be "Do linear algebra on n-dimensional arrays".

melissawm commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -122,146 +95,107 @@ can access this by using the syntax
-Since we are going to perform linear algebra operations on this data, it might
-be more interesting to have real numbers between 0 and 1 in each entry of the
-matrices to represent the RGB values. We can do that by setting
+To perform linear algebra, let's rescale these values into real numbers
+between 0 and 1:
```
In the original version, I had also written "perform linear algebra" and it was pointed out to me that it's not exactly a correct way to say it. BTW in math, at least in portuguese, we DO say "perform linear algebra" ;) So maybe "perform linear algebra operations" is better.

bjnath replied:
This sounds like one of those grammatical conundra @rossbar was warning about in howto-docs.

I don't understand the reasoning behind appending the word "operations" nor its relevance if that's not how mathematicians talk. We perform calculus, not calculus operations. 

If anyone wants to defend "operations" I'm open to learning something.

Another option is to can change "perform" to "do" and walk away.

stefanv replied:
I'd find it equally odd if you wrote "I'm about to perform calculus" and then proceeded to call `np.diff`. Just like you wouldn't typically say "I'm about to perform electrical engineering", or "I'm about to do some food science".

Linear algebra is a whole branch of mathematics: "one does not simply" apply it wholesale.

melissawm commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -122,146 +95,107 @@ can access this by using the syntax
+Methods from linear algebra can reduce data into a smaller approximate version
+that retains the most important features of the original. We'll do that using
+`SVD (Singular Value Decomposition)
+<https://blog.statsbot.co/singular-value-decomposition-tutorial-52c695315254?gi=2729cca75ba3>`_.
```
As I mentioned in #16244, I would propose keeping the wikipedia link for maintainability, even at the expense of clarity in this case.

bjnath replied:
The impermanence is a good point. I tried to get around this by saving the page in archive.org, but it didn't save the LaTeX. I made a screenshot of the article; we can link to that (giving credit, of course).  The Wikipedia article is so impenetrable to nonspecialists that it's like somebody's playing a joke.  If that's our only option I think we're better off leaving off the link.

bjnath replied:
I've found a nice explanation that's already on archive.org: https://web.archive.org/web/20200102142016/https://www.cs.princeton.edu/picasso/mats/PCA-Tutorial-Intuition_jp.pdf.

bjnath replied:
When we write a NumPy tutorial on writing NumPy tutorials, we should include @melissawm's point about choosing outside links carefully.

melissawm reviewed (commented):
With minor comments here and on #16244, this looks really good. Thanks, @bjnath !

bjnath commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -9,62 +9,38 @@ Tutorial: Linear algebra on n-dimensional arrays
+**Prerequisites**
 
-**Learning Objectives**
+Some linear algebra knowledge and basic familiarity with NumPy's ndararrays.
```
Yikes. Thanks.

bjnath commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -9,62 +9,38 @@ Tutorial: Linear algebra on n-dimensional arrays
-If you want to be able to run the examples in this tutorial, you should also
-have `matplotlib <https://matplotlib.org/>`_ and `SciPy <https://scipy.org>`_
-installed on your computer.
+You'll work with arrays of one, two, and three dimensions -- multiplying,
+scaling, transposing, and reordering them. You'll see NumPy broadcasting in
+action, and put some of NumPy's linear algebra tools to use. By the end,
+you'll be able to repeat the steps on an image of your own.
```
Standardize is certainly the goal. My thinking is that while teachers need to bear in mind categories like "Intended Student" and "Learning Objectives" so they can be sure to cover all the categories, it's not helpful to announce each one to the students as if the teaching had been mechanized.

The student picks up from the sentence that it's a Learning Objective, especially if we always make the second sentence a learning objective (which, further, is a standardization).

Still, signposts can be helpful. We can say things like "What we'll cover" as a livelier alternative (or leave it as "Learning objectives" if people prefer) and "What you'll learn."

But @rossbar  has his finger on it; what I'm hoping we can do is standardize our tutorial format.

bjnath replied:
@rossbar , I think we have a meeting point:

   **What we'll cover**
   **What you'll learn**
   **What you'll need**

bjnath commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -75,41 +51,38 @@ We can see the image using the `matplotlib.pyplot.imshow` function::
-   necessary to use the command ``plt.show()`` to show the image window. 
-		     
+   If you're using the IPython shell, type ``plt.show()`` if the image fails
+   to appear. You'll need `ctrl-C` to return to IPython.
```
Totally true. There's actually a bigger issue here. Procida would say the tutorial dictates the tools, so rather than "If you're using IPython," the tutorial says "Start IPython." 

I have mixed feelings. We'd like people using their favorite tools, but we don't want to have to explain every possible thing that can go wrong on all of them.

bjnath commented:
It was a great tutorial, Melissa -- the pace and exposition were perfect; it's clear you've been a teacher.

bjnath commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -122,146 +95,107 @@ can access this by using the syntax
-For example, from the theory, one might expect ``s`` and ``Vt`` to be
-compatible for multiplication. However, this is not true as ``s`` does not
-have a second axis. Executing
+Unexpectedly, ``s`` has just one dimension. It can't be multipled with ``Vt``.
```
Like this?
```
just one dimension. Under matrix algebra rules, it can't...
```

bjnath commented on doc/source/user/tutorial-svd.rst:
```diff
@@ -272,13 +206,10 @@ have a second axis. Executing
+results in a ``ValueError``.
+
+The one-dimensional ``s`` is an economized version of the diagonal matrix
+:math:`\Sigma` that we're looking for. We need to construct :math:`\Sigma` from ``s``:
```
Is this something Melissa should add or that I should put back?

bjnath commented:
I'm not sure we've converged enough for a new commit.

bjnath committed:
Merge branch 'tutorial-svd-edit' of github.com:bjnath/numpy-1 into tutorial-svd-edit

bjnath committed:
DOC: Incorporate some proposed changes to PR #16245

* New title
* Added prefatory tutorial headings
* Specified *matrix* multiply
* Clarified why k is 10 not 50
* Sidesteps "perform linear algebra"
* New durable link for SVD explanation

What isn't addressed:

* "You'll need ctrl-C to return to IPython'
(https://github.com/numpy/numpy/pull/16245#discussion_r429025078)

* https://github.com/numpy/numpy/pull/16245#discussion_r430106777

bjnath commented:
Pushed d8c309d to resolve most points above.

Awaiting resolution:
* "You'll need ctrl-C to return to IPython'     (https://github.com/numpy/numpy/pull/16245#discussion_r429025078)
* https://github.com/numpy/numpy/pull/16245#discussion_r430106777



![image](https://user-images.githubusercontent.com/6691888/82913876-d5150480-9f3c-11ea-8440-32cebef3af5e.png)

bjnath commented:
Continued in gh-16386.
